### PR TITLE
TST: Added test suite for dgamma distribution

### PR DESCRIPTION
**Pull request by MatteoRaso**
The distribution had absolutely no tests, which runs the risk of letting bad code get passed into the codebase.

#### Reference issue

No reference issue
#### What does this implement/fix?

Adds tests for dgamma, as there weren't any previously.

### CODE REVIEW

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+        assert_almost_equal(y_3, stats.dgamma.pdf(0.03, 0.01))
+
+    def test_frozen(self):
+        f = stats.dgamma(1.1)
+        assert_equal(f.pdf(1.3), stats.dgamma.pdf(1.3, 1.1))
+
+    def test_cdf(self):
+        vals = stats.dgamma.ppf([0.001, 0.5, 0.999], 1.3)
+        np.allclose([0.001, 0.5, 0.999], stats.dgamma.cdf(vals, 1.3))
+
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:3446]:
```suggestion
```
Covered by
https://github.com/scipy/scipy/blob/0406594b35bf0b382e06d2d56ed0ac1171fbe645/scipy/stats/tests/test_continuous_basic.py#L556

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+    def test_sf(self):
+        vals = stats.dgamma.ppf([0.001, 0.5, 0.999], 1.3)
+        S = stats.dgamma.sf(vals, 1.3)
+        assert_almost_equal(S, 1 - stats.dgamma.cdf(vals, 1.3))
+
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:3451]:
```suggestion
```
Covered by
https://github.com/scipy/scipy/blob/0406594b35bf0b382e06d2d56ed0ac1171fbe645/scipy/stats/tests/test_continuous_basic.py#L563

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+        #    return np.where(x > 0, 0.5 + fac, 0.5 - fac)
+
+        #def dgamma_isf(x, a):
+        #    return 1 / (1 - dgamma_cdf(x, a))
+
+        y_1 = 2.0033322961491538102693606074253344367238154381383
+        y_2 = 132.60693799743787520601808222040250683064041329549
+
+        np.isclose(stats.dgamma.isf(4.7, 0.01), y_1, atol = 1e-13)
+        np.isclose(stats.dgamma.isf(0.01, 1.01), y_2, atol = 1e-13)
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:3469]:
```suggestion
```
ISF is checked against SF
https://github.com/scipy/scipy/blob/0406594b35bf0b382e06d2d56ed0ac1171fbe645/scipy/stats/tests/test_continuous_basic.py#L563

**Review by mdhaber — CHANGES_REQUESTED** [review]
Thanks; just a few simplifications and we can merge.

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+
+    def test_pdf(self):
+        #Reference values calculated by hand using
+        #the defintion from the Scipy documentation
+        y_1 = 0.03790773263085006
+        assert_almost_equal(y_1, stats.dgamma.pdf(-1.3, 0.3))
+        y_2 = 0.10061666529975229
+        assert_almost_equal(y_2, stats.dgamma.pdf(2.7, 4.3))
+        y_3 = 0.15705884416519492
+        assert_almost_equal(y_3, stats.dgamma.pdf(0.03, 0.01))
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:3437]: Also, we don't use `assert_almost_equal` in new code.

image

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+class TestDgamma:
+    def test_logpdf(self):
+        x = np.array([1, 0.3, 4])
+        a = 1.3
+        y = stats.dgamma.pdf(x, a)
+        assert_allclose(y, np.exp(stats.dgamma.logpdf(x, a)))
+
+    def test_pdf(self):
+        #Reference values calculated by hand using
+        #the defintion from the Scipy documentation
```
[the remark this answers is not in the document]
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:3431]: I implemented the pdf using numpy, and then used that to give me my reference values.

File: scipy/stats/tests/test_distributions.py [hunk 6]
```diff
@@ -3419,6 +3419,54 @@ def test_isf(self):
...
+        y_1 = 0.03790773263085006
+        assert_almost_equal(y_1, stats.dgamma.pdf(-1.3, 0.3))
+        y_2 = 0.10061666529975229
+        assert_almost_equal(y_2, stats.dgamma.pdf(2.7, 4.3))
+        y_3 = 0.15705884416519492
+        assert_almost_equal(y_3, stats.dgamma.pdf(0.03, 0.01))
+
+    def test_frozen(self):
+        f = stats.dgamma(1.1)
+        assert_equal(f.pdf(1.3), stats.dgamma.pdf(1.3, 1.1))
```
[the remark this answers is not in the document]
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: Done.

**Comment by WarrenWeckesser** [comment]
The relation to the gamma distribution could be used to test the CDF also, with something like

    res = stats.dgamma.cdf(x, a)
    ref = np.where(x < 0, 0.5*stats.gamma.sf(-x, a),
                   0.5 + 0.5*stats.gamma.cdf(x, a))
    assert_allclose(res, ref, rtol=1e-12, atol=8e-15)

where `x` and `a` are generated as in `test_pdf`.  This snippet could be added to `test_pdf`, with the name changed to `test_pdf_cdf` (to save CPU time generating the random input), or it could be a separate test (to separate concerns of the tests and provide more focused results if the test fails).

A few experiments with `dgamma.cdf` and `dgamma.sf` exposed a weakness of the current implementation: precision is lost in the tails as a result of subtractive cancellation.  I'll create a new issue for that.

**Comment by mdhaber** [comment]
But since the CDF is already tested in test_continuous, let's not pull the author back and forth here. I'll open a separate issue about distribution testing standards. For now, @MatteoRaso, please just fix the lint error.

**Comment by WarrenWeckesser** [comment]
That's fine.  I'll take a closer look at the tests when I submit a fix for https://github.com/scipy/scipy/issues/17805

File: scipy/stats/tests/test_distributions.py [hunk 7]
```diff
@@ -3419,6 +3419,19 @@ def test_isf(self):
         assert np.isclose(stats.gamma.isf(1e-50, 100),
                           330.6557590436547, atol=1e-13)
 
+class TestDgamma:
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:3422]:
```suggestion

class TestDgamma:
```

**Review by mdhaber — APPROVED** [review]
Hopefully, the most recent commit fixes the lint issue. I'll merge when that passes (since `[skip ci]` doesn't seem to work when the commit is added via GitHub...).

**Comment by mdhaber** [comment]
Thanks @MatteoRaso!

**Comment by MatteoRaso** [comment]
No problem.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Pull request merged**